Introduce Kiali validations to Istio Config plugin page

lucasponce opened the pull request:
This PR is longer than expected, but it incorporates some important changes:
- The main feature is to improve the Service Mesh -> Istio Config page by adding Kiali validations (@hhovsepy's work).
- In the Istio Config page, the Kubernetes resources are fetched by the OpenShift Console, and the Kiali validations are queried using the Kiali API through the ConsolePlugin proxy.
- The ConsolePlugin needs to provide the information of the internal Kiali Service, and the operator is updated to support this information; the operator uses the same auto-discovery technique used by the Kiali public Url.
- This design is different from the used in the iframes; in this case, the network calls to the Kiali API happen from the same domain of the OpenShift Console, so it has a different approach (it's the expected way for plugins to invoke third party APIs).

@hhovsepy @jmazzitelli, this PR is now ready for review; I've tested it both in developer mode and in an OpenShift cluster.
I've manually tested the molecule tests from the operator.

hhovsepy commented:
@lucasponce it works as expected.

lucasponce commented:
Thanks, @hhovsepy, I have a pending minor refactor for naming, but I think this one will be shortly ready for review.

lucasponce commented:
Ouch, this PR has introduced a bad bug in the OpenShift Console. Still investigating.

lucasponce commented:
I've promoted this PR for reviewing, I'm still doing some tests, but it can be reviewed.

hhovsepy reviewed (approved):
LGTM, working as expected.

lucasponce commented:
It has a problem in the cluster; it's not ready yet for merging.
I'll keep reviewing.

hhovsepy reviewed (approved):
Verified:
![Screenshot from 2022-08-24 12-14-29](https://user-images.githubusercontent.com/604313/186393327-c90e8d26-bde6-45d1-9b2f-6670621d3777.png)

lucasponce commented:
Thanks, @hhovsepy; glad you could test it.

@jmazzitelli, do you have any comments on the operator changes?

jmazzitelli commented:
i'm reviewing now.

jmazzitelli commented:
This is failing for me. I think you will need to rebase because this doesn't work on 4.11 (and we fixed that in a previous PR merge - so I think a rebase will fix it). My issue:

```
      message: 'pods "ossmplugin-77cc579754-mhxkd" is forbidden: unable to validate
        against any security context constraint: [pod.metadata.annotations.container.seccomp.security.alpha.kubernetes.io/ossmplugin:
        Forbidden: seccomp may not be set provider "anyuid": Forbidden: not usable
        by user or serviceaccount provider "nonroot": Forbidden: not usable by user
        or serviceaccount provider "bookinfo-scc": Forbidden: not usable by user or
        serviceaccount provider "n1-scc": Forbidden: not usable by user or serviceaccount
        provider "topology-generator-scc": Forbidden: not usable by user or serviceaccount
        provider "travel-scc": Forbidden: not usable by user or serviceaccount provider
        "hostmount-anyuid": Forbidden: not usable by user or serviceaccount provider
        "machine-api-termination-handler": Forbidden: not usable by user or serviceaccount
        provider "hostnetwork": Forbidden: not usable by user or serviceaccount provider
        "hostaccess": Forbidden: not usable by user or serviceaccount provider "node-exporter":
        Forbidden: not usable by user or serviceaccount provider "privileged": Forbidden:
        not usable by user or serviceaccount]'
```

jmazzitelli reviewed (changes requested):
@lucasponce rebase this with the latest main branch. This breaks when I try it on 4.11, but I think it might already be fixed in the latest main.

lucasponce commented:
I can rebase, but I've also noticed that kiali (and probably this one) doesn't deploy in 4.10 out of the box.
Perhaps unrelated, but just to bring to the attention that perhaps we may miss something in <4.11 clusters.

I'll rebase in a moment.

jmazzitelli commented:
oh, you know what? I got my clusters confused. That error above came from a 4.10 cluster. Hmmm... what is going on?

lucasponce commented:
Rebased.
I've installed a CRC with 4.11, and the PR looks ok from here.
Some demo apps (the "alpha" and "beta" and the "traffic-generator" on the "bookinfo") have problems on my side, but it's unrelated to this work.

jmazzitelli commented:
> Some demo apps (the "alpha" and "beta" and the "traffic-generator" on the "bookinfo") have problems on my side, but it's unrelated to this work.

right.. @josunect is taking care of that work.

jmazzitelli commented:
>  I've also noticed that kiali (and probably this one) doesn't deploy in 4.10 out of the box.

I'm in the process of fixing that now. See the gory details here: https://github.com/kiali/kiali/issues/5409

jmazzitelli commented:
now things should work if you rebase (again) on main

lucasponce commented:
Rebased and tested in 4.10.
![image](https://user-images.githubusercontent.com/1662329/186856032-be929b61-5cde-4772-af46-e561c1b9db25.png)
![image](https://user-images.githubusercontent.com/1662329/186856097-60c344cd-ec11-49f6-8d0c-501d4aa6c891.png)

If there are no blockers, I suggest consolidating this and adding additional changes in the operator and/or code in the follow-up commit.

Otherwise, this one can get stuck in side aspects.

Note that @hhovsepy will continue these main efforts on the detail pages.

josunect commented:
> > Some demo apps (the "alpha" and "beta" and the "traffic-generator" on the "bookinfo") have problems on my side, but it's unrelated to this work.
> 
> right.. @josunect is taking care of that work.

Yes, I've sent a PR, https://github.com/kiali/kiali/pull/5397 or the issue https://github.com/kiali/kiali/issues/5394, in case you want to check the details.

jmazzitelli commented:
I don't see those new settings going into the config map `plugin-config.json`. They are in the CR. I see this:

```
$ oc get ossmplugin ossmplugin -n ossmplugin -o jsonpath={.status.kiali} | jq
{
  "serviceName": "kiali",
  "serviceNamespace": "istio-system",
  "servicePort": "20001",
  "url": "https://kiali-istio-system.apps.ocp4.local"
}
```

But the configmap has:

```
$ oc get cm -n ossmplugin plugin-conf -o yaml | head -n 7
apiVersion: v1
data:
  plugin-config.json: |
    {
      "kialiUrl": "https://kiali-istio-system.apps.ocp4.local"
    }
kind: ConfigMap
```

So I question if the plugin is actually getting those service config setttings.

jmazzitelli commented:
This PR needs to pass these settings to the configmap template:
https://github.com/kiali/openshift-servicemesh-plugin/blob/main/operator/roles/default/ossmplugin-deploy/templates/openshift/configmap-plugin.yaml#L10

lucasponce commented:
> This PR needs to pass these settings to the configmap template:

No. Those settings are needed only for the ConsolePlugin artifact.

jmazzitelli commented:
> No. Those settings are needed only for the OpenShift ConsolePlugin resource.

Right! I just noticed that. I updated my comment and deleted that last one.

lucasponce commented:
> Right! I just noticed that. I updated my comment and deleted that last one.

@hhovsepy can give more details, but in short, updating the ConsolePlugin, allows that these settings are a "constant" for the UI part.

For development mode, this setting is passed in the package.json

jmazzitelli reviewed (approved):
I see the Configuration column now.

I tested on both 4.10 and 4.11 - works on both platforms.

jmazzitelli commented:
> @hhovsepy can give more details, but in short, updating the ConsolePlugin, allows that these settings are a "constant" for the UI part.

Right - but don't forget, this means this does not support OSSM multi-tenancy... so we still have this to worry about:
* https://github.com/kiali/openshift-servicemesh-plugin/issues/28
